sharktank/sharktank/layers/paged_llama_attention_block.py

```diff
@@ -0,0 +1,265 @@
+# Copyright 2024 Advanced Micro Devices, Inc
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from typing import Optional
+
+import math
+
+import torch
+import torch.nn.functional as F
+
+from .base import Theta, ThetaLayer
+from .linear import LinearLayer
+from .norm import RMSNormLayer
+from .rotary_embedding import RotaryEmbeddingLayer
+from .kv_cache import PagedKVCache
+
+__all__ = [
+    "PagedLlamaAttentionBlock",
+]
+
+
+class PagedLlamaAttentionBlock(ThetaLayer):
+    """Implements a self attention layer in the style of Llama using a
+    paged cache."""
+
+    def __init__(
+        self,
+        theta: Theta,
+        *,
+        block_index: int,
+        cache: PagedKVCache,
+        head_count: int,
+        head_dim: int,
+        head_count_kv: int,
+        rms_epsilon: float,
+        use_hf: bool = False,
+    ):
+        super().__init__(theta)
+        self.add_module(
+            "attn_norm", RMSNormLayer(theta("attn_norm"), epsilon=rms_epsilon)
+        )
+        self.add_module("attn_q", LinearLayer(theta("attn_q")))
+        self.add_module("attn_k", LinearLayer(theta("attn_k")))
+        self.add_module("attn_v", LinearLayer(theta("attn_v")))
+        self.add_module("attn_output", LinearLayer(theta("attn_output")))
+        self.add_module(
```

Comment:
`attn_output_norm` is a Grok specific layer, will break llama models.

Reply:
And this is why it fails the test!

Reply:
I think we can just add it as its own layer in the new grok model, and I'll just remove it here